### PR TITLE
Support `tuple`, `struct` types in `kevm-pyk`, `abi.md`

### PR DESCRIPTION
**Pull request by palinatolmach**
The KEVM part (`kevm-pyk`, `abi.md`) of the PR introducing support for `struct`, `tuple` types:  https://github.com/runtimeverification/evm-semantics/pull/2051.

Quoting the original PR, user-defined structs are tuples under the hood (Solidity [docs](https://docs.soliditylang.org/en/latest/abi-spec.html#handling-tuple-types)) and that they are not tightly packed (two uint8 are taking 2 words in the calldata), so we can just recursively unpack the types from the ABI.

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kevm.py [hunk 1]
```diff
@@ -344,6 +344,10 @@ def abi_bool(b: KInner) -> KApply:
     def abi_type(type: str, value: KInner) -> KApply:
         return KApply('abi_type_' + type, [value])
 
+    @staticmethod
+    def abi_tuple(values: Sequence[KInner]) -> KApply:
+        return KApply('abi_type_tuple', values)
```
Comment (ehildenb) [inline kevm-pyk/src/kevm_pyk/kevm.py:349]: For generality and matchnig the signature of `KApply` (and other methods here), it should probably use `Iterable`:

```suggestion
    @staticmethod
    def abi_tuple(values: Iterable[KInner]) -> KApply:
        return KApply('abi_type_tuple', values)
```

https://github.com/runtimeverification/pyk/blob/79c45d9f1ae355b591cbd2e8b77c6205ab170835/src/pyk/kast/inner.py#L368

File: kevm-pyk/src/kevm_pyk/kevm.py [hunk 2]
```diff
@@ -388,12 +392,21 @@ def intlist(ints: list[KInner]) -> KApply:
...
     @staticmethod
-    def typed_args(args: list[KInner]) -> KApply:
-        res = KApply('.List{"_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs"}_TypedArgs')
-        for i in reversed(args):
-            res = KApply('_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs', [i, res])
+    def typed_args(args: list[KApply], res: KApply | None = None) -> KApply:
+        res = KEVM.empty_typedargs() if res is None else res
+        for arg in reversed(args):
+            res = KEVM.to_typed_arg(arg, res)
         return res
```
Comment (ehildenb) [inline kevm-pyk/src/kevm_pyk/kevm.py:399]: Can `build_cons` be used here? https://github.com/runtimeverification/pyk/blob/79c45d9f1ae355b591cbd2e8b77c6205ab170835/src/pyk/kast/inner.py#L699

```suggestion
    def typed_args(args: list[KApply], res: KApply | None = None) -> KApply:
        res = KEVM.empty_typedargs() if res is None else res
        return build_cons(res, '_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs', reversed(args))
```

Does it produce the correct results?

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 1 above)
Comment (palinatolmach) [reply kevm-pyk/src/kevm_pyk/kevm.py]: Thanks! Fixed in https://github.com/runtimeverification/evm-semantics/pull/2105/commits/06b2b44d379d6a57878aad3f8d23a30ce390cd4a.

File: kevm-pyk/src/kevm_pyk/kevm.py (hunk 2 above)
Comment (palinatolmach) [reply kevm-pyk/src/kevm_pyk/kevm.py:399]: It doesn't produce the same result, and a previously passing test containing structs now fails if `build_cons` is used. I'm still investigating why, will post an update here.

Comment (palinatolmach) [reply kevm-pyk/src/kevm_pyk/kevm.py:399]: Update: I'm using the following test for investigation:
```solidity
contract StructTypeTest {
    struct Vars {
        uint8 a;
        uint32 b;
    }
    ...
    function test_vars(Vars calldata vars) public view {
       ...
    }
}
```

 In `definition.kore`, there's a difference in the rule for `test_vars`:
```
`method_StructTypeTest_S2KtestZUndvars_uint8_uint32`(V0_a,V1_b))=>
`#abiCallData(_,_)_EVM-ABI_Bytes_String_TypedArgs`(#token("\"test_vars\"","String"),
```
depending on whether the original `to_typed_arg` or `build_cons` is used. 

For `to_typed_arg`, which behaves correctly on proofs, it is: 
```
`_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs`(`abi_type_tuple`(
   `_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs`(`abi_type_uint8`(V0_a),
   `_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs`(`abi_type_uint32`(V1_b),
   `.List{"_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs"}_TypedArgs`(.KList)))),
`.List{"_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs"}_TypedArgs`(.KList)))
```

While for `build_cons`, it is:
```
`_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs`(`abi_type_tuple`(
   `_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs`(`abi_type_uint8`(V0_a),
   `.List{"_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs"}_TypedArgs`(.KList))),
 `.List{"_,__EVM-ABI_TypedArgs_TypedArg_TypedArgs"}_TypedArgs`(.KList))
```

So, in short, with `build_cons` we're losing the second item in the tuple (`Vars.b`). But I can look into refactoring original `typed_args(...)` and  `to_typed_arg (...)` further.

Comment (palinatolmach) [reply kevm-pyk/src/kevm_pyk/kevm.py:403]: @anvacaru @tothtamas28 can I ask for your opinion on whether we should expect [build_cons](https://github.com/runtimeverification/pyk/blob/79c45d9f1ae355b591cbd2e8b77c6205ab170835/src/pyk/kast/inner.py#L699) to process a `KApply` with several `args` successfully? The `tuple` in question is: 
```
KApply(label=KLabel(name='abi_type_tuple', params=()),args=
(KApply(label=KLabel(name='abi_type_uint8', params=()), args=(KVariable(name='V0_a', sort=None),)), 
 KApply(label=KLabel(name='abi_type_uint32', params=()), args=(KVariable(name='V1_b', sort=None),))))
```
Please see the previous comment for the description of the issue: only the first element of the tuple (`V0_a`) gets added to the `test_vars` signature in the corresponding rule, where `test_vars` is:
```solidity
    struct Vars {
        uint8 a;
        uint32 b;
    }
    ...
    function test_vars(Vars calldata vars) public view {
```

Comment (tothtamas28) [reply kevm-pyk/src/kevm_pyk/kevm.py:403]: `build_cons` seems like an adequate helper here. Consider setting up a small unit test that compares the two implementations on a few inputs, that'll help detecting any potential differences.

**Comment by palinatolmach** [comment]
Closing as these changes are included in https://github.com/runtimeverification/evm-semantics/pull/2136.

**Pull request closed without merging**